File: src/github.ts
// Minting an installation token and dispatching a workflow. No SDK: two requests.

const API = "https://api.github.com"
const UA = "jshvn-dispatch"

/** A refusal from GitHub, carrying the status so the caller can decide whether to retry. */
export class GitHubError extends Error {
  readonly status: number
  constructor(status: number, message: string) {
    super(message)
    this.status = status
  }
}

/**
 * GitHub refused the request itself -- a missing workflow file, an App without
 * `actions: write`, a ref that does not exist. Retrying cannot change the answer, so the
 * step should fail now and say so rather than spend three attempts on a typo. 429 and 408
 * are the two 4xx that do clear on their own.
 */
export const isFatal = (err: unknown): boolean =>
  err instanceof GitHubError &&
  err.status >= 400 &&
  err.status < 500 &&
  ![408, 429].includes(err.status)

const b64url = (bytes: ArrayBuffer): string => {
  let s = ""
  for (const b of new Uint8Array(bytes)) s += String.fromCharCode(b)
  return btoa(s).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "")
}

const b64urlJson = (value: unknown): string =>
  b64url(new TextEncoder().encode(JSON.stringify(value)).buffer as ArrayBuffer)

/**
 * GitHub issues App keys as PKCS#1 ("BEGIN RSA PRIVATE KEY"); WebCrypto only imports
 * PKCS#8. Convert once before storing the secret:
 *   openssl pkcs8 -topk8 -nocrypt -in app.private-key.pem -out app.pkcs8.pem
 */
const importKey = (pkcs8Pem: string): Promise<CryptoKey> => {
  if (pkcs8Pem.includes("BEGIN RSA PRIVATE KEY"))
    throw new Error("private key is PKCS#1; convert it with `openssl pkcs8 -topk8 -nocrypt`")
  const body = pkcs8Pem.replace(/-----[A-Z ]+-----/g, "").replace(/\s+/g, "")
  const der = Uint8Array.from(atob(body), (c) => c.charCodeAt(0))
  return crypto.subtle.importKey(
    "pkcs8",
    der.buffer as ArrayBuffer,
    { name: "RSASSA-PKCS1-v1_5", hash: "SHA-256" },
    false,
    ["sign"],
  )
}

/** A short-lived App JWT. Backdated 60s because GitHub rejects a future `iat`. */
export const appJwt = async (
  appId: string,
  pkcs8Pem: string,
  now = Date.now(),
): Promise<string> => {
  const iat = Math.floor(now / 1000) - 60
  const signed = `${b64urlJson({ alg: "RS256", typ: "JWT" })}.${b64urlJson({ iat, exp: iat + 540, iss: appId })}`
  const sig = await crypto.subtle.sign(
    "RSASSA-PKCS1-v1_5",
    await importKey(pkcs8Pem),
    new TextEncoder().encode(signed),
  )
  return `${signed}.${b64url(sig)}`
}

const headers = (auth: string) => ({
  authorization: `Bearer ${auth}`,
  accept: "application/vnd.github+json",
  "x-github-api-version": "2022-11-28",
  "user-agent": UA,
})

/** Exchange an App JWT for an installation token. Valid one hour; never persisted. */
export const installationToken = async (
  jwt: string,
  installationId: string,
  fetch: typeof globalThis.fetch = globalThis.fetch,
): Promise<string> => {
  const res = await fetch(`${API}/app/installations/${installationId}/access_tokens`, {
    method: "POST",
    headers: headers(jwt),
  })
  if (!res.ok)
    throw new GitHubError(res.status, `installation token: ${res.status} ${await res.text()}`)
  const body = (await res.json()) as { token?: string }
  if (!body.token) throw new Error("installation token: response carried no token")
  return body.token
}

/** POST a workflow_dispatch. GitHub answers 204 with no body, so there is no run id here. */
export const dispatchWorkflow = async (
  token: string,
  target: { repo: string; workflow: string; ref?: string; inputs?: Record<string, string> },
  fetch: typeof globalThis.fetch = globalThis.fetch,
): Promise<void> => {
  const url = `${API}/repos/${target.repo}/actions/workflows/${target.workflow}/dispatches`
  const res = await fetch(url, {
    method: "POST",
    headers: { ...headers(token), "content-type": "application/json" },
    body: JSON.stringify({ ref: target.ref ?? "main", inputs: target.inputs ?? {} }),
  })
  if (res.status !== 204)
    throw new GitHubError(
      res.status,
      `dispatch ${target.repo}/${target.workflow}: ${res.status} ${await res.text()}`,
    )
}

File: github.go
package main

// The GitHub App's side: sign a JWT, find the App's installation on the org, mint an
// installation token, dispatch a workflow. No SDK, three requests.

import (
	"bytes"
	"context"
	"crypto"
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha256"
	"crypto/x509"
	"encoding/base64"
	"encoding/json"
	"encoding/pem"
	"errors"
	"fmt"
	"io"
	"net"
	"net/http"
	"time"
)

// backoff is the wait before each retry of a transient failure: three retries, 70 s in all.
// A retry whose wait would pass the context's deadline is not made.
var backoff = []time.Duration{10 * time.Second, 20 * time.Second, 40 * time.Second}

// StatusError is GitHub answering with something other than 2xx.
type StatusError struct {
	Status int
	Msg    string
}

func (e *StatusError) Error() string { return e.Msg }

// retryable is a failure another attempt may clear. Any other 4xx is a refusal no retry can
// change: a missing workflow file, an App without the permission, no installation.
//
// A request that starts a run (safe false) is retried only when GitHub cannot have acted on
// it: the connection never opened, or GitHub turned it away unread (408, 429). A 5xx or a
// timeout can follow a dispatch GitHub already accepted, and a retry would be a second run.
func retryable(err error, safe bool) bool {
	var s *StatusError
	if errors.As(err, &s) {
		return s.Status == 408 || s.Status == 429 || s.Status >= 500 && safe
	}
	var op *net.OpError
	return safe || errors.As(err, &op) && op.Op == "dial"
}

// ParseKey reads the App's private key. GitHub issues PKCS#1; PKCS#8 is accepted too, so a
// converted key still works.
func ParseKey(data []byte) (*rsa.PrivateKey, error) {
	block, _ := pem.Decode(data)
	if block == nil {
		return nil, errors.New("private key: no PEM block")
	}
	if k, err := x509.ParsePKCS1PrivateKey(block.Bytes); err == nil {
		return k, nil
	}
	k, err := x509.ParsePKCS8PrivateKey(block.Bytes)
	if err != nil {
		return nil, fmt.Errorf("private key: neither PKCS#1 nor PKCS#8: %w", err)
	}
	rk, ok := k.(*rsa.PrivateKey)
	if !ok {
		return nil, fmt.Errorf("private key: %T, not RSA", k)
	}
	return rk, nil
}

// AppJWT signs a short-lived App JWT. Backdated 60 s because GitHub rejects a future iat,
// and 9 minutes long against GitHub's 10-minute ceiling.
func AppJWT(appID string, key *rsa.PrivateKey, now time.Time) (string, error) {
	enc := base64.RawURLEncoding
	iat := now.Unix() - 60
	claims, err := json.Marshal(map[string]any{"iat": iat, "exp": iat + 540, "iss": appID})
	if err != nil {
		return "", err
	}
	signed := enc.EncodeToString([]byte(`{"alg":"RS256","typ":"JWT"}`)) + "." + enc.EncodeToString(claims)
	sum := sha256.Sum256([]byte(signed))
	sig, err := rsa.SignPKCS1v15(rand.Reader, key, crypto.SHA256, sum[:])
	if err != nil {
		return "", err
	}
	return signed + "." + enc.EncodeToString(sig), nil
}

// GitHub dispatches workflows as the App. Prepare mints one installation token and keeps it
// in memory only.
type GitHub struct {
	API   string // https://api.github.com
	AppID string
	Org   string // the one installation this App has
	Key   *rsa.PrivateKey
	HTTP  *http.Client
	Sleep func(time.Duration)
	Now   func() time.Time

	token string
}

// Prepare mints the installation token, once per process.
func (g *GitHub) Prepare(ctx context.Context) error {
	if g.token != "" {
		return nil
	}
	tok, err := g.mint(ctx)
	if err != nil {
		return fmt.Errorf("installation token: %w", err)
	}
	g.token = tok
	return nil
}

func (g *GitHub) Dispatch(ctx context.Context, repo, file string) error {
	if err := g.Prepare(ctx); err != nil {
		return err
	}
	_, err := g.call(ctx, "POST", "/repos/"+repo+"/actions/workflows/"+file+"/dispatches", g.token,
		[]byte(`{"ref":"main","inputs":{}}`), false)
	return err
}

func (g *GitHub) mint(ctx context.Context) (string, error) {
	jwt, err := AppJWT(g.AppID, g.Key, g.Now())
	if err != nil {
		return "", err
	}
	b, err := g.call(ctx, "GET", "/orgs/"+g.Org+"/installation", jwt, nil, true)
	if err != nil {
		return "", err
	}
	var inst struct{ ID int64 }
	if err := json.Unmarshal(b, &inst); err != nil || inst.ID == 0 {
		return "", fmt.Errorf("installation for %s: no id in %.200s", g.Org, b)
	}
	b, err = g.call(ctx, "POST", fmt.Sprintf("/app/installations/%d/access_tokens", inst.ID), jwt, nil, true)
	if err != nil {
		return "", err
	}
	var tok struct{ Token string }
	if err := json.Unmarshal(b, &tok); err != nil || tok.Token == "" {
		return "", errors.New("access token: none in the response")
	}
	return tok.Token, nil
}

// call sends one request, retrying transient failures after each backoff step. safe says a
// repeat has no effect beyond the first; see retryable.
func (g *GitHub) call(ctx context.Context, method, path, auth string, body []byte, safe bool) ([]byte, error) {
	for attempt := 0; ; attempt++ {
		b, err := g.once(ctx, method, path, auth, body)
		if err == nil || !retryable(err, safe) || attempt == len(backoff) || ctx.Err() != nil {
			return b, err
		}
		if dl, ok := ctx.Deadline(); ok && g.Now().Add(backoff[attempt]).After(dl) {
			return b, err
		}
		g.Sleep(backoff[attempt])
	}
}

func (g *GitHub) once(ctx context.Context, method, path, auth string, body []byte) ([]byte, error) {
	req, err := http.NewRequestWithContext(ctx, method, g.API+path, bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "Bearer "+auth)
	req.Header.Set("Accept", "application/vnd.github+json")
	req.Header.Set("X-GitHub-Api-Version", "2022-11-28")
	req.Header.Set("User-Agent", "katoptra-dispatch")
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	res, err := g.HTTP.Do(req)
	if err != nil {
		return nil, err
	}
	defer res.Body.Close()
	b, err := io.ReadAll(io.LimitReader(res.Body, 1<<20))
	if err != nil {
		return nil, err
	}
	if res.StatusCode/100 != 2 {
		return nil, &StatusError{res.StatusCode, fmt.Sprintf("%s %s: %d %.500s", method, path, res.StatusCode, b)}
	}
	return b, nil
}
